File: 278_first_bad_version.js
/**
 * Definition for isBadVersion()
 *
 * @param {integer} version number
 * @return {boolean} whether the version is bad
 * isBadVersion = function(version) {
 *     ...
 * };
 */

/**
 * @param {function} isBadVersion()
 * @return {function}
 */
var solution = function(isBadVersion) {
    /**
     * @param {integer} n Total versions
     * @return {integer} The first bad version
     */
    return function(n) {
        if (n === 1){
            return 1;
        }

        let left = 1;
        let right = n;

        while (left<= right){
            const middle = Math.floor((left+right)/2);

            if (isBadVersion(middle) || isBadVersion(left)){
                right = middle;
            } else if (isBadVersion(right)){
                left = middle+1;
            }
            if (left === right){
                right--;
            }
        }
        return left;
    };
};

//Time Complexity O(log n), Space Complexity O(1);

File: 278_first_bad_version.py
# The isBadVersion API is already defined for you.
# def isBadVersion(version: int) -> bool:
import math

class Solution:
    def firstBadVersion(self, n: int) -> int:
        if n==1:
            return 1
        
        left = 1
        right = n

        while left <= right:
            middle = math.floor((left+right)/2)

            if isBadVersion(middle) or isBadVersion(left):
                right = middle
            elif isBadVersion(right):
                left = middle+1
            if left == right:
                right-=1
        return left
